src/cpu_shader/path_count.rs

```diff
@@ -0,0 +1,157 @@
+// Copyright 2023 The Vello authors
+// SPDX-License-Identifier: Apache-2.0 OR MIT OR Unlicense
+
+use vello_encoding::{BumpAllocators, LineSoup, Path, SegmentCount, Tile};
+
+use crate::cpu_dispatch::CpuBinding;
+
+use super::util::{span, Vec2};
+
+const TILE_SCALE: f32 = 1.0 / 16.0;
+
+fn path_count_main(
+    bump: &mut BumpAllocators,
+    lines: &[LineSoup],
+    paths: &[Path],
+    tile: &mut [Tile],
+    seg_counts: &mut [SegmentCount],
+) {
+    for line_ix in 0..bump.lines {
+        let line = lines[line_ix as usize];
+        let p0 = Vec2::from_array(line.p0);
+        let p1 = Vec2::from_array(line.p1);
+        let is_down = p1.y >= p0.y;
+        let (xy0, xy1) = if is_down { (p0, p1) } else { (p1, p0) };
+        let s0 = xy0 * TILE_SCALE;
+        let s1 = xy1 * TILE_SCALE;
+        let count = span(s0.x, s1.x) + span(s0.y, s1.y) - 1;
+
+        let dx = (s1.x - s0.x).abs();
+        let dy = s1.y - s0.y;
+        if dx + dy == 0.0 {
+            continue;
+        }
+        if dy == 0.0 && s0.y.floor() == s0.y {
+            continue;
+        }
+        let dy_dxdy = dy / (dx + dy);
+        let a = 1.0 - dy_dxdy;
+        let is_positive_slope = s1.x >= s0.x;
+        let sign = if is_positive_slope { 1.0 } else { -1.0 };
+        let xt0 = (s0.x * sign).floor();
+        let c = s0.x * sign - xt0;
+        let y0 = s0.y.floor();
+        let ytop = if s0.y == s1.y { s0.y.ceil() } else { y0 + 1.0 };
+        let b = dy_dxdy * c + a * (ytop - s0.y);
```

Comment:
Is there a reason for the different formulations between the WGSL and Rust versions (`dy_dxdy`/`idxdy`, `a`, `b`, etc)? Should they be the same?

I'm also slightly confused about `a`. It's defined as `1.0 - dy / (dx + dy)` here but as `dx / (dx + dy)` in WGSL. It's then used in the same calculations below. Are these equivalent?

Reply:
Ok, this question is incredibly deep water. The short answer is that you're looking at an experiment to improve numerical robustness. The *ultimate* solution to the artifact was to discard horizontal line segments aligned to grid boundaries before it gets to this point. What happened there is a = 1 - epsilon, b = 0, while you need a = 1 for conservative line rasterization to generate the right answer. In CPU land, you can use a division instead of multiplying by reciprocal to make that case work, and use the fact that dx/(dx + dy) + dy/(dx + dy) = 1 to avoid having to compute two divisions. In GPU land, WGSL doesn't give you that guarantee.

What I think is the best path going forward is to undo that experiment and make CPU match GPU. It's possible (especially with carefully constructed adversarial examples) that you might trigger incorrect behavior - in fact, this is one of the things I'll look at for #373. But you're absolutely right that CPU and GPU should match unless there's a compelling reason otherwise.